daemon/lib/source/DobbyManager.cpp

```diff
@@ -1735,6 +1738,287 @@ bool DobbyManager::wakeupContainer(int32_t cd)
+        
+        if (setns(fdMountNs, CLONE_NEWNS) < 0)
+        {
+            AI_LOG_WARN("failed to setns to mount namespace for container with descriptor %d(%d - %m)", cd, errno);
+            close(fdMnt);
+            close(fdUserNs);
+            close(fdMountNs);
+            _exit(EXIT_FAILURE);
+        }
+        AI_LOG_INFO("setns succeeded for mount namespace for container with descriptor %d", cd);
+        
+        close(fdMountNs);
+        close(fdUserNs);
+
+        // switch to local user
+        setuid(0);
+        setgid(0);
+        if (mkdir(destination.c_str(), 0755) != 0)
+        {
+            AI_LOG_WARN("failed to create destination directory %s for container with descriptor %d", destination.c_str(), cd);
+            close(fdMnt);
+            _exit(EXIT_FAILURE);
+        }
+
+        AI_LOG_INFO("mkdir succeeded for directory %s on container with descriptor: %d", destination.c_str(), cd);
+
+        int ret = syscall(SYS_move_mount, fdMnt, "",  -EBADF, destination.c_str(), MOVE_MOUNT_F_EMPTY_PATH);
+        if (ret < 0)
+        {
+            AI_LOG_WARN("move_mount failed errno: %d container: %d", errno, cd);
+            close(fdMnt);
+            _exit(EXIT_FAILURE);
+        }
+
+        close(fdMnt);
+        _exit(EXIT_SUCCESS);
+    }
+
+    // wait for the forked child to finish
+    int status;
+    if ((waitpid(pid, &status, 0) == -1) || (WEXITSTATUS(status) != EXIT_SUCCESS))
+    {
+        AI_LOG_ERROR_EXIT("mount operation failed for container with descriptor %d", cd);
+        AI_LOG_FN_EXIT();
+        return false;
+    }
+
+    AI_LOG_INFO("%s is mounted inside %s", source.c_str(),  id.c_str());
+    AI_LOG_FN_EXIT();
+    return true;
+}
+
+// -----------------------------------------------------------------------------
+/**
+ *  @brief removes a mount from a running container
+ *
+ *  @param[in]  cd           The descriptor of the container to checkpoint.
+ *  @param[in]  source
+ *
+ *  @return true if a container was successfully restored.
+ */
+bool DobbyManager::removeMount(int32_t cd, const std::string &source)
+{
+     AI_LOG_FN_ENTRY();
+
+    std::lock_guard<std::mutex> locker(mLock);
+
+    // find the container
+    auto it = mContainers.cbegin();
+    for (; it != mContainers.cend(); ++it)
+    {
+        if (it->second && (it->second->descriptor == cd))
+            break;
+    }
+
+    if (it == mContainers.cend())
+    {
+        AI_LOG_WARN("failed to find container with descriptor %d", cd);
+        AI_LOG_FN_EXIT();
+        return false;
+    }
+
+    const ContainerId &id = it->first;
+    
+    const pid_t containerPid = it->second->containerPid;
+    if (containerPid == 0)
+    {
+        AI_LOG_WARN("failed to find container pid for container with descriptor %d", cd);
+        AI_LOG_FN_EXIT();
+        return false;
+    }
+    
+    AI_LOG_INFO("containerPid: %d", containerPid);
+
+    // fork the process
+    pid_t pid = fork();
```

Comment:
Same as above, I think you can avoid the `fork` call by switching the effective uid / gid on the thread running in the mount namespace.

Reply:
I've removed the fork() and replaced it with setresgid/uid and  callinnamespace() calls. It works!